pkg/pillar/cmd/tpmmgr/handlecertificate.go

```diff
@@ -0,0 +1,384 @@
+	}
+	// create soft Ecdh Certificate
+	return createEveNodeCertificateSoft(ctx, certType)
+}
+
+// create the certificate on Tpm
+func createEveNodeCertificateOnTpm(ctx *tpmMgrContext,
+	certType evecommon.ZCertType, deviceCert *x509.Certificate) error {
+
+	switch certType {
+	// of ECDH Type
+	case evecommon.ZCertType_Z_CERT_TYPE_DEVICE_ECDH_EXCHANGE:
+		return createEveNodeEcdhCerteOnTpm(ctx, deviceCert)
+
+	// TBD:XXX create attest certificate, others, if any
+	default:
+		errStr := fmt.Sprintf("Certificate type: invalid")
+		log.Errorf(errStr)
+		return errors.New(errStr)
+	}
+}
+
+// create the ecdh template
+func createEcdhSoftTemplate() x509.Certificate {
+	serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
+	serialNumber, _ := rand.Int(rand.Reader, serialNumberLimit)
+	notBefore := time.Now()
+	// twenty years
+	notAfter := notBefore.AddDate(20, 0, 0)
+
+	template := x509.Certificate{
+		SerialNumber: serialNumber,
+		Subject: pkix.Name{
+			Country:      []string{"US"},
+			Province:     []string{"CA"},
+			Locality:     []string{"Santa Clara"},
+			Organization: []string{"Zededa, Inc"},
+			CommonName:   "Cipher Block",
+		},
+		NotBefore: notBefore,
+		NotAfter:  notAfter,
+
+		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		BasicConstraintsValid: true,
+	}
+	return template
+}
+
+func createEveNodeCertificateSoft(ctx *tpmMgrContext,
+	certType evecommon.ZCertType) error {
+
+	/// not ECDH type return
+	if certType != evecommon.ZCertType_Z_CERT_TYPE_DEVICE_ECDH_EXCHANGE {
+		errStr := fmt.Sprintf("not ecdh certificate type")
+		log.Errorf(errStr)
+		return errors.New(errStr)
+	}
+
+	// create the ecdh certificate
+	priv, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if err != nil {
+		errStr := fmt.Sprintf("curve get fail, %v", err)
+		log.Errorf(errStr)
+		return errors.New(errStr)
+	}
+	publicKey := priv.PublicKey
+	template := createEcdhSoftTemplate()
+
+	derBytes, err := x509.CreateCertificate(rand.Reader,
+		&template, &template, publicKey, priv)
+	if err != nil {
+		errStr := fmt.Sprintf("certificate create fail, %v", err)
+		log.Errorf(errStr)
+		return errors.New(errStr)
+	}
+
+	certBlock := &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: derBytes,
+	}
+
+	certBytes := pem.EncodeToMemory(certBlock)
+	if certBytes == nil {
+		errStr := fmt.Sprintf("PEM Encode error: empty bytes")
+		log.Errorf(errStr)
+		return errors.New(errStr)
+	}
+
+	privBytes, err := x509.MarshalPKCS8PrivateKey(priv)
+	if err != nil {
+		errStr := fmt.Sprintf("certificate marshal fail, %v", err)
+		log.Errorf(errStr)
+		return errors.New(errStr)
+	}
+
+	keyBlock := &pem.Block{
+		Type:  "PRIVATE KEY",
+		Bytes: privBytes,
+	}
+
+	keyBytes := pem.EncodeToMemory(keyBlock)
+	if keyBytes == nil {
+		errStr := fmt.Sprintf("PEM Encode error: empty bytes")
+		log.Errorf(errStr)
+		return errors.New(errStr)
+	}
```

Comment:
This private key should be written to a file in /config presumably /config/ecdh.key.pem and not passed around in pubsub.

Note that the cert needs to be signed by the device key.